keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1116 @@
+be overridden by any of the ANPs in a cluster. The aforementioned tooling will not 
+be a primary development goal during the alpha version of this API, and will most 
+likely be completed during the beta development cycle. 
+
+### Future Work
+
+Although the scope of the AdminNetworkPolicies is extensive, the above proposal
+intends to only solve the documented use cases. However, we would
+also like to consider the following set of proposals as future work items:
+- **Audit Logging**: Very often cluster administrators want to log every connection
+  that is either denied or allowed by a firewall rule and send the details to
+  an IDS or any custom tool for further processing of that information.
+  With the introduction of `deny` rules, it may make sense to incorporate the
+  cluster-scoped policy resources with a new field, say `auditPolicy`, to
+  determine whether a connection matching a particular rule/policy must be
+  logged or not.
+
+## Design Details
+
+### AdminNetworkPolicy API Design
+
+The following new `AdminNetworkPolicy` API will be added to the `policy.networking.k8s.io` 
+API group.  
+
+```golang
+type AdminNetworkPolicy struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  // Specification of the desired behavior of AdminNetworkPolicy.
+  Spec     AdminNetworkPolicySpec
+  // Status to be reported by the CNI, this is not standardized in alpha and 
+  // CNIs should report what they see fit in relation to their AdminNetworkPolicy
+  // implementation
+  // +optional 
+  Status   []metav1.Condition
+}
+
+type AdminNetworkPolicySpec struct {
+  // Int value bound to 0 - 999, with 0 signaling the special "baseline" priority
+  Priority      Int 
+  // Allows for multiple AdminNetworkPolicy rules at a single priority level
+  // If empty no rules are applied
+  // A total of 100 total rules will be allowed per each network policy instance, 
+  // this rule count will be calculated as the total summation of the 
+  // AdminNetworkPolicyRules, AdminNetworkPolicyIngressRules, and AdminNetworkPolicyEgressRules
+  // in a single AdminNetworkPolicy Instance. 
+  PolicyRules   []AdminNetworkPolicyRule
+}
+
+type AdminNetworkPolicyRule struct {
+  // No implicit isolation of AppliedTo Pods/Namespaces. 
+  // If no rules are a defined no behavior will be observed 
+  AppliedTo   AppliedTo
+  Ingress     []AdminNetworkPolicyIngressRule
+  Egress      []AdminNetworkPolicyEgressRule
+}
+
+// Exactly one of the selector pointers should be set in a given AppliedTo
+type AppliedTo struct {
+  NamespaceSelector         *metav1.LabelSelector
+  NamespaceAndPodSelector   *NamespaceAndPodSelector
+}
+
+// NamespaceAndPodSelector allows the user to select a namespace/s and optionally 
+// select a given set of pods in that namespace/s
+type NamespaceAndPodSelector struct {
+  NamespaceSelector   metav1.LabelSelector
+  // Used to explicitly select pods within a namespace
+  PodSelector         *metav1.LabelSelector
```

Comment:
maybe is to give semantics to empty selector vs nil?
 but that is one of the complains to current NetworkPolicy API , that is not very nice UX

Reply:
I believe this was simply a mistake... If the user decides to utilize a `NamespaceAndPodSelector` struct they should always be specifying a Namespace and set of pods, otherwise they would simply use a vanilla `NamespaceSelector`

Reply:
We should look at convention - it seems like it's always a pointer in other APIs.

Reply:
Well in Netpol at least the `Subject` podSelector _isn't_ a pointer

Reply:
I followed this ^ convention

Reply:
Let's press ahead and get a consult at API PR time,